src/pybind/mgr/test_orchestrator/module.py

```diff
@@ -0,0 +1,253 @@
+        self.log.info("wait: completions={0}".format(completions))
+
+        incomplete = False
+
+        # Our `wait` implementation is very simple because everything's
+        # just an API call.
+        for c in completions:
+            if not isinstance(c, TestReadCompletion) and \
+                    not isinstance(c, TestWriteCompletion):
+                raise TypeError(
+                    "wait() requires list of completions, not {0}".format(
+                        c.__class__
+                    ))
+
+            if c.is_complete:
+                continue
+
+            if not c.is_read:
+                self._progress("update", c.id, c.message, 0.5)
+
+            try:
+                c.execute()
+            except Exception as e:
+                self.log.exception("Completion {0} threw an exception:".format(
+                    c.message
+                ))
+                c.error = e
+                c._complete = True
+                if not c.is_read:
+                    self._progress("complete", c.id)
+            else:
+                if c.is_complete:
+                    if not c.is_read:
+                        self._progress("complete", c.id)
+
+            if not c.is_complete:
+                incomplete = True
+
+        return not incomplete
+
+    def available(self):
+        return True, ""
+
+    def __init__(self, *args, **kwargs):
+        super(TestOrchestrator, self).__init__(*args, **kwargs)
+
+        self._initialized = threading.Event()
+        self._shutdown = threading.Event()
+
+    def shutdown(self):
+        self._shutdown.set()
+
+    def serve(self):
+
+        self._initialized.set()
+
+        while not self._shutdown.is_set():
+            # XXX hack (or is it?) to kick all completions periodically,
+            # in case we had a caller that wait()'ed on them long enough
+            # to get persistence but not long enough to get completion
+
+            global all_completions
+            self.wait(all_completions)
+            all_completions = [c for c in all_completions if not c.is_complete]
+
+            self._shutdown.wait(5)
+
+    @deferred_read
+    def get_inventory(self, node_filter=None):
+        """
+        There is no guarantee which devices are returned by get_inventory.
+        """
+        try:
+            c_v_out = check_output(['ceph_volume', 'inventory', '--format', 'json'])
+        except OSError:
+            cmd = """
+            . /tmp/ceph-volume-virtualenv/bin/activate
+            ceph-volume inventory --format json
+            """
+            c_v_out = check_output(cmd, shell=True)
+
+        for out in c_v_out.splitlines():
+            if not out.startswith(b'-->') and not out.startswith(b' stderr'):
+                self.log.error(out)
+                devs = []
+                for device in json.loads(out):
+                    dev = orchestrator.InventoryDevice()
+                    if 'partitions' not in device["sys_api"]:
```

Comment:
This filters our partitions in the JSON output. Relates to #25390. Can be removed afterwards

Reply:
#25390 is merged now, so I guess this can be removed?